File: assignments/assignment-2-solution/src-csharp/TeaShop.UnitTest/UserInterface/UserInterfaceTests.cs
using Assignment2Solution.Domain.Inventory;
using Assignment2Solution.UserInterface.PaymentMethodGenerator;
using Assignment2Solution.UserInterface.Query;

namespace TeaShop.UnitTest.UserInterface;

public class UserInterfaceTests
{
    [Fact]
    public void QueryBuilder_Build_ReturnsConfiguredQuery()
    {
        // Arrange
        var repository = new InventoryRepository();
        // Simulate user input for all prompts
        var inputLines = new[]
        {
            "Green", // Name contains
            "Y", // Is available
            "10", // Price min
            "20", // Price max
            "4", // Rating min
            "5", // Rating max
            "A", // Sort by Price,
            "D" // Sort by Star rating
        };
        var input = new StringReader(string.Join(Environment.NewLine, inputLines));
        var output = new StringWriter();
        var builder = new QueryBuilder(repository, input, output);

        // Act
        var query = builder.Build();

        // Assert
        Assert.NotNull(query);
        var filters = query.AppliedFiltersAndSorts;
        Assert.Contains(filters, f => f.Contains("Green"));
        Assert.Contains(filters, f => f.Contains("In Stock"));
        Assert.Contains(filters, f => f.Contains("$10.00") && f.Contains("$20.00"));
        Assert.Contains(filters, f => f.Contains("Star rating between 4") && f.Contains("5"));
        Assert.Contains(filters, f => f.Contains("Sort: Price (ascending)"));
        Assert.Contains(filters, f => f.Contains("Sort: Star rating (descending)"));
    }

    [Fact]
    public void ApplePayPaymentMethod_CreateStrategy_ReturnsStrategy()
    {
        // Arrange
        var method = new ApplePayPaymentStrategyGenerator();
        var input = new StringReader("user@example.com" + Environment.NewLine);
        var output = new StringWriter();

        // Act
        var strategy = method.CreateStrategy(input, output);

        // Assert
        Assert.NotNull(strategy);
        Assert.Equal("Apple Pay", method.Name);

        // Check output prompt
        Assert.Contains("Enter Apple Username:", output.ToString());
    }

    [Fact]
    public void CreditCardPaymentMethod_CreateStrategy_ReturnsStrategy()
    {
        // Arrange
        var method = new CreditCardPaymentStrategyGenerator();
        var input = new StringReader("1234567890123456" + Environment.NewLine);
        var output = new StringWriter();

        // Act
        var strategy = method.CreateStrategy(input, output);

        // Assert
        Assert.NotNull(strategy);
        Assert.Equal("Credit Card", method.Name);
    }

    [Fact]
    public void CryptoCurrencyPaymentMethod_CreateStrategy_ReturnsStrategy()
    {
        // Arrange
        var method = new CryptoCurrencyPaymentStrategyGenerator();
        var input = new StringReader("0x123" + Environment.NewLine + "sig123" + Environment.NewLine);
        var output = new StringWriter();

        // Act
        var strategy = method.CreateStrategy(input, output);

        // Assert
        Assert.NotNull(strategy);
        Assert.Equal("CryptoCurrency", method.Name);
    }
}

File: assignments/assignment-2-solution/src-java/src/test/java/teashop/unittest/userinterface/UserInterfaceTests.java
package teashop.unittest.userinterface;

import assignment2solution.domain.inventory.InventoryRepository;
import assignment2solution.userinterface.paymentbuilder.ApplePayPaymentBuilder;
import assignment2solution.userinterface.paymentbuilder.CreditCardPaymentBuilder;
import assignment2solution.userinterface.paymentbuilder.CryptoCurrencyPaymentBuilder;
import assignment2solution.userinterface.querybuilder.InventoryQueryBuilder;

import org.junit.jupiter.api.Test;

import java.io.StringReader;
import java.io.StringWriter;

import static org.junit.jupiter.api.Assertions.*;

public class UserInterfaceTests {
    @Test
    public void queryBuilderBuildReturnsConfiguredQuery() {
        var repository = new InventoryRepository();
        var inputLines = new String[]{
            "Green",
            "Y",
            "10",
            "20",
            "4",
            "5",
            "A",
            "D"
        };
        var input = new StringReader(String.join(System.lineSeparator(), inputLines));
        var output = new StringWriter();
        var builder = new InventoryQueryBuilder(repository, input, output);

        var query = builder.build();

        assertNotNull(query);
        var filters = query.getAppliedFiltersAndSorts();
        assertTrue(filters.stream().anyMatch(f -> f.contains("Green")));
        assertTrue(filters.stream().anyMatch(f -> f.contains("In Stock")));
        assertTrue(filters.stream().anyMatch(f -> f.contains("$10.00") && f.contains("$20.00")));
        assertTrue(filters.stream().anyMatch(f -> f.contains("Star rating between 4") && f.contains("5")));
        assertTrue(filters.stream().anyMatch(f -> f.contains("Sort: Price (ascending)")));
        assertTrue(filters.stream().anyMatch(f -> f.contains("Sort: Star rating (descending)")));
    }

    @Test
    public void applePayPaymentMethodCreateStrategyReturnsStrategy() {
        var method = new ApplePayPaymentBuilder();
        var input = new StringReader("user@example.com" + System.lineSeparator());
        var output = new StringWriter();

        var strategy = method.createStrategy(input, output);

        assertNotNull(strategy);
        assertEquals("Apple Pay", method.getName());
        assertTrue(output.toString().contains("Enter Apple Username:"));
    }

    @Test
    public void creditCardPaymentMethodCreateStrategyReturnsStrategy() {
        var method = new CreditCardPaymentBuilder();
        var input = new StringReader("1234567890123456" + System.lineSeparator());
        var output = new StringWriter();

        var strategy = method.createStrategy(input, output);

        assertNotNull(strategy);
        assertEquals("Credit Card", method.getName());
    }

    @Test
    public void cryptoCurrencyPaymentMethodCreateStrategyReturnsStrategy() {
        var method = new CryptoCurrencyPaymentBuilder();
        var input = new StringReader("0x123" + System.lineSeparator() + "sig123" + System.lineSeparator());
        var output = new StringWriter();

        var strategy = method.createStrategy(input, output);

        assertNotNull(strategy);
        assertEquals("CryptoCurrency", method.getName());
    }
}
